### PR TITLE
Add one to Mul typeclass and add prod() to prelude.

### PR DESCRIPTION
**Pull request by duvenaud**
This isn't part of any demo, I was just idly trying to flesh out the typeclass hierarchy.

However, I'm not totally sure this is the right way to go forward.  My next step was going to be to add an identity element to ```VSpace``` to make it match the mathematical definition.  But I'm not sure if I should try to share ```Mul```'s  ```one```, and have them both inherit from a new typeclass called ```has_identity``` or some such.

If anyone reading this has thought about architecting these sorts of typeclass hierarchies before, please chime in!

### CODE REVIEW

File: prelude.dx [hunk 1]
```diff
@@ -279,8 +280,7 @@ def scan' (init:a) (body:n->a->a) : n=>a = snd $ scan init \i x. dup (body i x)
 -- TODO: allow tables-via-lambda and get rid of this
 def fsum (xs:n->Real) : Real = snd $ withAccum \ref. for i. ref += xs i
 def sum  (_: Add v) ?=> (xs:n=>v) : v = reduce zero (+) xs
--- TODO: Add 'one' to Mul typeclass when adts branch is merged, then add prod
--- def prod (_: Mul v) ?=> (xs:n=>v) : v = reduce one  (*) xs
+def prod (_: Mul v) ?=> (xs:n=>v) : v = reduce one  (*) xs
```
Comment (oxinabox) [inline prelude.dx:283]: Trivial suggestion to remove weird spacing
```suggestion
def prod (_: Mul v) ?=> (xs:n=>v) : v = reduce one (*) xs
```

**Comment by dougalm** [comment]
> add an identity element to VSpace

You mean an identity for the scalar field or the vector? The vector already has one because it inherits from `Add`. For now, the scalar is hard-coded as `Real` rather than a type parameter, so it has one too. A two-parameter type class would be a better fit but I suspect we'll have to tweak the dictionary synthesis a bit to get that to work.

**Comment by oxinabox** [comment]
> However, I'm not totally sure this is the right way to go forward. My next step was going to be to add an identity element to VSpace to make it match the mathematical definition. But I'm not sure if I should try to share Mul's one, and have them both inherit from a new typeclass called has_identity or some such.


I still haven't caught up on the type-class stuff.
But a vector space doesn't have a multiplicative identity.
It has an attached scalar field and that field has a multiplicative identity that is the vectors scalar multiplicative identity.
(which is a bit of a taugology)

**Comment by duvenaud** [comment]
Yes, I was thinking about the ultra-general case where we parameterize the scalar field too, but now that I think about it, that's probably overkill.

**Comment by apaszke** [comment]
Hmmm I thought that the idea of having type classes such as `Mul` and `Add` was simply to allow overloading of `*` and `+` without necessarily implying any structure of those operations? For example, if I had a type of non-empty words, then it's easy to define `+` as concatenation, but it does not have an identity element in that type, because it's just a semigroup. IMO we should avoid bundling too much in those basic type classes that simply expose overloading, and then define semantically charged type-classes (such as assuming that an operation has an identity element) later. WDYT?

**Comment by oxinabox** [comment]
Strings are a a free moniod.
Its unclear to me how far one can go (practically) with type-classes representing abstract algebraic things.
I suspect Dex could take this quite far.
Which would be really cool.
`(+, zero)` and `*(, 1)` are nice moniods for many types.

```
julia> "abc" * "xyx"
"abcxyx"

julia> one(String)
""

julia> "abc" * one(String)
"abc"

julia> "abc"^3
"abcabcabc"

julia> "abc" ^ 0
""
```

You can do the same thing with `+` instead of `*` and `*` instead of `^`, and `zero` instead of `one`.
but some have argued that is less elegant, since `+` is normally commutivative.

**Pull request merged**